File: app/src/main/java/com/fieldservice/app/presentation/ticketdetails/TicketDetailsViewModel.java
package com.fieldservice.app.presentation.ticketdetails;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.fieldservice.app.domain.model.Ticket;
import com.fieldservice.app.domain.model.TicketStatus;
import com.fieldservice.app.domain.repository.TicketRepository;
import com.fieldservice.app.presentation.UiState;

import java.util.List;

/**
 * Expõe o chamado selecionado (derivado da fonte reativa do repositório, de modo que
 * avançar o status atualize a tela sozinho) e o estado de "atualizando".
 */
public class TicketDetailsViewModel extends ViewModel {

    private final TicketRepository ticketRepository;
    private final String ticketId;

    private final LiveData<UiState<Ticket>> uiState;
    private final MutableLiveData<Boolean> isUpdating = new MutableLiveData<>(false);

    public TicketDetailsViewModel(TicketRepository ticketRepository, String ticketId) {
        this.ticketRepository = ticketRepository;
        this.ticketId = ticketId;
        this.uiState = Transformations.map(ticketRepository.observeTickets(), this::findState);
    }

    private UiState<Ticket> findState(List<Ticket> tickets) {
        for (Ticket ticket : tickets) {
            if (ticket.getId().equals(ticketId)) {
                return UiState.success(ticket);
            }
        }
        return UiState.empty();
    }

    public LiveData<UiState<Ticket>> getUiState() {
        return uiState;
    }

    public LiveData<Boolean> getIsUpdating() {
        return isUpdating;
    }

    /** Move o chamado para o próximo status do fluxo (aceitar, deslocar, chegar, atender, finalizar). */
    public void advance(TicketStatus nextStatus) {
        isUpdating.setValue(true);
        ticketRepository.updateTicketStatus(ticketId, nextStatus, () -> isUpdating.setValue(false));
    }
}

File: legacy-kotlin/app/src/main/java/com/fieldservice/app/presentation/ticketdetails/TicketDetailsViewModel.kt
package com.fieldservice.app.presentation.ticketdetails

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.fieldservice.app.domain.model.Ticket
import com.fieldservice.app.domain.repository.TicketRepository
import com.fieldservice.app.presentation.UiState
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class TicketDetailsViewModel(
    private val ticketRepository: TicketRepository,
    private val ticketId: String
) : ViewModel() {

    val uiState: StateFlow<UiState<Ticket>> = ticketRepository.observeTickets()
        .map { tickets -> tickets.find { it.id == ticketId } }
        .map { ticket -> if (ticket != null) UiState.Success(ticket) else UiState.Empty }
        .catch { error -> emit(UiState.Error(error.message ?: "Não foi possível carregar o chamado")) }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), UiState.Loading)

    private val _isAccepting = MutableStateFlow(false)
    val isAccepting: StateFlow<Boolean> = _isAccepting

    fun acceptTicket() {
        viewModelScope.launch {
            _isAccepting.value = true
            ticketRepository.acceptTicket(ticketId)
            _isAccepting.value = false
        }
    }
}
